server/admin/tree_gc_test.go

```diff
@@ -175,6 +177,176 @@ func TestDeletedTreeGC_RunOnce(t *testing.T) {
+
+		// wantCount is the count of successfully deleted trees.
+		wantCount int
+		// wantErrs defines which strings must be present in the resulting error.
+		wantErrs []string
+	}{
+		{
+			desc:        "snapshotErr",
+			snapshotErr: errors.New("snapshot err"),
+			wantErrs:    []string{"snapshot err"},
+		},
+		{
+			desc:     "listErr",
+			listErr:  errors.New("list err"),
+			wantErrs: []string{"list err"},
+		},
+		{
+			desc:              "snapshotCommitErr",
+			trees:             []*trillian.Tree{logTree1, logTree2, mapTree},
+			snapshotCommitErr: errors.New("commit err"),
+			wantErrs:          []string{"commit err"},
+		},
+		{
+			desc:        "beginErr",
+			trees:       []*trillian.Tree{logTree1, logTree2},
+			beginErrs:   []error{errors.New("begin err"), nil},
+			deleteErrs:  []error{nil, nil},
+			commitErrs:  []error{nil, nil},
+			wantTreeIDs: []int64{logTree1.TreeId, logTree2.TreeId},
+			wantCount:   1,
+			wantErrs:    []string{"begin err"},
+		},
+		{
+			desc:        "deleteErr",
+			trees:       []*trillian.Tree{logTree1, logTree2},
+			beginErrs:   []error{nil, nil},
+			deleteErrs:  []error{errors.New("cannot delete logTree1"), nil},
+			commitErrs:  []error{nil, nil},
+			wantTreeIDs: []int64{logTree1.TreeId, logTree2.TreeId},
+			wantCount:   1,
+			wantErrs:    []string{"cannot delete logTree1"},
+		},
+		{
+			desc:        "commitErr",
+			trees:       []*trillian.Tree{logTree1, logTree2},
+			beginErrs:   []error{nil, nil},
+			deleteErrs:  []error{nil, nil},
+			commitErrs:  []error{errors.New("commit err"), nil},
+			wantTreeIDs: []int64{logTree1.TreeId, logTree2.TreeId},
+			wantCount:   1,
+			wantErrs:    []string{"commit err"},
+		},
+		{
+			// logTree1 = delete successful
+			// logTree2 = delete error
+			// mapTree  = commit error
+			// badTS    = timestamp parse error (no HardDeleteTree() call)
+			desc:        "multipleErrors",
+			trees:       []*trillian.Tree{logTree1, logTree2, mapTree, badTS},
+			beginErrs:   []error{nil, nil, nil},
+			deleteErrs:  []error{nil, errors.New("delete err"), nil},
+			commitErrs:  []error{nil, nil, errors.New("commit err")},
+			wantTreeIDs: []int64{logTree1.TreeId, logTree2.TreeId, mapTree.TreeId},
+			wantCount:   1,
+			wantErrs:    []string{"delete err", "commit err", "error parsing delete_time"},
+		},
+	}
+
+	ctx := context.Background()
+	for _, test := range tests {
+		as := storage.NewMockAdminStorage(ctrl)
+
+		listTX := storage.NewMockReadOnlyAdminTX(ctrl)
+		lastTXCall := as.EXPECT().Snapshot(gomock.Any()).Return(listTX, test.snapshotErr)
+		listTX.EXPECT().ListTrees(gomock.Any(), true /* includeDeleted */).AnyTimes().Return(test.trees, test.listErr)
+		listTX.EXPECT().Close().AnyTimes().Return(nil)
+		listTX.EXPECT().Commit().AnyTimes().Return(test.snapshotCommitErr)
+
+		// Sanity check test setup
+		if l1, l2, l3, l4 := len(test.beginErrs), len(test.deleteErrs), len(test.commitErrs), len(test.wantTreeIDs); l1 != l2 || l1 != l3 || l1 != l4 {
```

Comment:
Not worth fixing for a test, but this self-consistency check hints you might be better off having a `[]testOp` with `testOp` being a struct holding a begin/delete/commit error and a treeID, so it's impossible for a maintainer to create a new invalid testcase.

Reply:
Good call. I went ahead and did this, as it makes it a lot easier to read the expected hard-delete flow.